File: 0545-boundary-of-binary-tree/0545-boundary-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isLeaf(TreeNode* node){
        if(node->left == NULL && node->right == NULL) return true;
        return false;
    }
    void addLeaves(vector<int>& ans, TreeNode* node){
        if(!node) return;
        if(isLeaf(node)) {
            ans.push_back(node->val);
        }
        else{
            if(node->left) addLeaves(ans, node->left);
            if(node->right) addLeaves(ans, node->right);
        }
    }
    vector<int> boundaryOfBinaryTree(TreeNode* root) {
        vector<int> ans;
        if(!root) return ans;
        if(!isLeaf(root)) ans.push_back(root->val);
        TreeNode* t = root->left;
        while(t != NULL){
            if(!isLeaf(t)) ans.push_back(t->val);
            if(t->left) t = t->left;
            else t = t->right;
        }

        addLeaves(ans, root);

        stack<int> stack;
        t = root->right;

        while(t != NULL){
            if(!isLeaf(t)) stack.push(t->val);

            if(t->right != NULL) t = t->right;
            else t = t->left;
        }

        while(!stack.empty()){
            ans.push_back(stack.top());
            stack.pop();
        }

        return ans;

    }
};

File: 0545-boundary-of-binary-tree/0545-boundary-of-binary-tree.py

class Solution(object):
    def isLeaf(self, t):
        return t.left == None and t.right == None

    def addLeaves(self, res, root):
        if self.isLeaf(root):
            res.append(root.val)
        else:
            if root.left != None:
                self.addLeaves(res, root.left)
            if root.right != None:
                self.addLeaves(res, root.right)

    def boundaryOfBinaryTree(self, root):
        if root == None:
            return []

        res = []
        if not self.isLeaf(root):
            res.append(root.val)

        # Add left boundary (excluding leaves)
        t = root.left
        while t != None:
            if not self.isLeaf(t):
                res.append(t.val)
            if t.left != None:
                t = t.left
            else:
                t = t.right

        # Add all leaf nodes
        self.addLeaves(res, root)

        # Add right boundary (excluding leaves) in reverse order
        stack = []
        t = root.right
        while t is not None:
            if not self.isLeaf(t):
                stack.append(t.val)
            if t.right != None:
                t = t.right
            else:
                t = t.left

        while stack:
            res.append(stack.pop())

        return res